File: apps/mobile/lib/core/models/enums.dart
import 'package:flutter/material.dart';

/// Domain enums — the Dart mirror of `@jardimja/shared`
/// (`packages/shared/src/enums.ts`). The wire value of every enum matches the
/// SCREAMING_SNAKE string used by the backend so JSON round-trips without a
/// translation layer. pt-BR labels mirror `ServiceTypeLabel` etc. exactly.

/// Shared helper: resolve an enum from its wire value with a fallback.
T _fromWire<T>(
  List<T> values,
  String Function(T) wireOf,
  String? wire,
  T fallback,
) {
  if (wire == null) return fallback;
  for (final v in values) {
    if (wireOf(v) == wire) return v;
  }
  return fallback;
}

// ── ServiceType ────────────────────────────────────────────────────────────
/// Catalogue of bookable services. `OUTRO` is the free-text escape hatch.
enum ServiceType {
  corteGrama('CORTE_GRAMA', 'Cortar grama', Icons.grass),
  poda('PODA', 'Poda', Icons.content_cut),
  paisagismo('PAISAGISMO', 'Paisagismo', Icons.deck),
  limpeza('LIMPEZA', 'Limpeza', Icons.cleaning_services),
  retiradaFolhas('RETIRADA_FOLHAS', 'Retirada de folhas', Icons.air),
  adubacao('ADUBACAO', 'Adubação', Icons.compost),
  plantio('PLANTIO', 'Plantio', Icons.local_florist),
  controlePragas('CONTROLE_PRAGAS', 'Controle de pragas', Icons.pest_control),
  irrigacao('IRRIGACAO', 'Sistema de irrigação', Icons.water_drop),
  jardimCompleto('JARDIM_COMPLETO', 'Jardim completo', Icons.yard),
  outro('OUTRO', 'Outro', Icons.more_horiz);

  const ServiceType(this.wire, this.label, this.icon);

  final String wire;
  final String label;
  final IconData icon;

  static ServiceType fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, ServiceType.outro);
}

// ── JobStatus ──────────────────────────────────────────────────────────────
/// Uber-style state machine for a service request.
enum JobStatus {
  draft('DRAFT', 'Rascunho'),
  analyzing('ANALYZING', 'Analisando'),
  quoted('QUOTED', 'Orçado'),
  matching('MATCHING', 'No marketplace'),
  offered('OFFERED', 'Com propostas'),
  accepted('ACCEPTED', 'Aceito'),
  scheduled('SCHEDULED', 'Agendado'),
  enroute('ENROUTE', 'A caminho'),
  arrived('ARRIVED', 'Chegou'),
  inProgress('IN_PROGRESS', 'Em andamento'),
  completed('COMPLETED', 'Concluído'),
  approved('APPROVED', 'Aprovado'),
  paid('PAID', 'Pago'),
  reviewed('REVIEWED', 'Avaliado'),
  cancelled('CANCELLED', 'Cancelado'),
  disputed('DISPUTED', 'Em disputa');

  const JobStatus(this.wire, this.label);

  final String wire;
  final String label;

  static JobStatus fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, JobStatus.draft);
}

// ── OfferStatus ────────────────────────────────────────────────────────────
enum OfferStatus {
  pending('PENDING', 'Pendente'),
  acceptedByGardener('ACCEPTED_BY_GARDENER', 'Aceitou o preço'),
  countered('COUNTERED', 'Contraproposta'),
  declined('DECLINED', 'Recusada'),
  chosen('CHOSEN', 'Escolhida'),
  expired('EXPIRED', 'Expirada'),
  withdrawn('WITHDRAWN', 'Retirada');

  const OfferStatus(this.wire, this.label);

  final String wire;
  final String label;

  static OfferStatus fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, OfferStatus.pending);
}

// ── PaymentStatus / PaymentMethod ──────────────────────────────────────────
enum PaymentStatus {
  pending('PENDING'),
  authorized('AUTHORIZED'),
  captured('CAPTURED'),
  split('SPLIT'),
  refunded('REFUNDED'),
  failed('FAILED');

  const PaymentStatus(this.wire);
  final String wire;

  static PaymentStatus fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, PaymentStatus.pending);
}

enum PaymentMethod {
  pix('PIX', 'PIX'),
  creditCard('CREDIT_CARD', 'Cartão de crédito'),
  googlePay('GOOGLE_PAY', 'Google Pay'),
  applePay('APPLE_PAY', 'Apple Pay');

  const PaymentMethod(this.wire, this.label);
  final String wire;
  final String label;

  static PaymentMethod fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, PaymentMethod.pix);
}

// ── UserRole ───────────────────────────────────────────────────────────────
enum UserRole {
  client('CLIENT'),
  gardener('GARDENER'),
  admin('ADMIN'),
  support('SUPPORT');

  const UserRole(this.wire);
  final String wire;

  static UserRole fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, UserRole.client);
}

// ── Equipment ──────────────────────────────────────────────────────────────
enum Equipment {
  rocadeira('ROCADEIRA', 'Roçadeira', Icons.agriculture),
  cortadorGrama('CORTADOR_GRAMA', 'Cortador de grama', Icons.grass),
  motosserra('MOTOSSERRA', 'Motosserra', Icons.carpenter),
  soprador('SOPRADOR', 'Soprador', Icons.air),
  triturador('TRITURADOR', 'Triturador', Icons.blender),
  escada('ESCADA', 'Escada', Icons.stairs),
  caminhao('CAMINHAO', 'Caminhão', Icons.local_shipping),
  pulverizador('PULVERIZADOR', 'Pulverizador', Icons.sanitizer),
  podadorAltura('PODADOR_ALTURA', 'Podador de altura', Icons.height);

  const Equipment(this.wire, this.label, this.icon);

  final String wire;
  final String label;
  final IconData icon;

  static Equipment fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, Equipment.rocadeira);
}

// ── DifficultyLevel ────────────────────────────────────────────────────────
enum DifficultyLevel {
  low('LOW', 'Baixa', Color(0xFF16A34A)),
  medium('MEDIUM', 'Média', Color(0xFFF59E0B)),
  high('HIGH', 'Alta', Color(0xFFEA580C)),
  extreme('EXTREME', 'Extrema', Color(0xFFDC2626));

  const DifficultyLevel(this.wire, this.label, this.color);

  final String wire;
  final String label;
  final Color color;

  static DifficultyLevel fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, DifficultyLevel.medium);
}

// ── UrgencyLevel ───────────────────────────────────────────────────────────
enum UrgencyLevel {
  flexible('FLEXIBLE', 'Flexível'),
  normal('NORMAL', 'Normal'),
  urgent('URGENT', 'Urgente'),
  emergency('EMERGENCY', 'Emergência');

  const UrgencyLevel(this.wire, this.label);

  final String wire;
  final String label;

  static UrgencyLevel fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, UrgencyLevel.normal);
}

// ── RiskLevel ──────────────────────────────────────────────────────────────
enum RiskLevel {
  low('LOW', 'Baixo'),
  medium('MEDIUM', 'Médio'),
  high('HIGH', 'Alto');

  const RiskLevel(this.wire, this.label);
  final String wire;
  final String label;

  static RiskLevel fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, RiskLevel.low);
}

// ── TerrainSlope ───────────────────────────────────────────────────────────
enum TerrainSlope {
  flat('FLAT', 'Plano'),
  gentle('GENTLE', 'Leve'),
  moderate('MODERATE', 'Moderado'),
  steep('STEEP', 'Íngreme');

  const TerrainSlope(this.wire, this.label);
  final String wire;
  final String label;

  static TerrainSlope fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, TerrainSlope.flat);
}

// ── AccessDifficulty ───────────────────────────────────────────────────────
enum AccessDifficulty {
  easy('EASY', 'Fácil'),
  moderate('MODERATE', 'Moderado'),
  hard('HARD', 'Difícil');

  const AccessDifficulty(this.wire, this.label);
  final String wire;
  final String label;

  static AccessDifficulty fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, AccessDifficulty.easy);
}

// ── MediaKind ──────────────────────────────────────────────────────────────
enum MediaKind {
  photo('PHOTO'),
  video('VIDEO'),
  audio('AUDIO'),
  document('DOCUMENT');

  const MediaKind(this.wire);
  final String wire;

  static MediaKind fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, MediaKind.photo);
}

// ── MessageKind ────────────────────────────────────────────────────────────
enum MessageKind {
  text('TEXT'),
  photo('PHOTO'),
  video('VIDEO'),
  audio('AUDIO'),
  location('LOCATION'),
  document('DOCUMENT'),
  system('SYSTEM');

  const MessageKind(this.wire);
  final String wire;

  static MessageKind fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, MessageKind.text);
}

// ── GardenerStatus ─────────────────────────────────────────────────────────
enum GardenerStatus {
  pendingVerification('PENDING_VERIFICATION', 'Em verificação'),
  active('ACTIVE', 'Ativo'),
  suspended('SUSPENDED', 'Suspenso'),
  rejected('REJECTED', 'Recusado');

  const GardenerStatus(this.wire, this.label);
  final String wire;
  final String label;

  static GardenerStatus fromWire(String? wire) =>
      _fromWire(values, (e) => e.wire, wire, GardenerStatus.pendingVerification);
}

File: packages/shared/src/enums.ts
/**
 * Domain enums — the shared vocabulary of JardimJá.
 *
 * These are declared as `const` objects (plus a union type) rather than TS
 * `enum`s so they can be used interchangeably as runtime values, Zod enum
 * inputs, Prisma enum mirrors, and JSON payloads without transpilation quirks.
 */

/** Catalogue of bookable services. `OUTRO` is the free-text escape hatch. */
export const ServiceType = {
  CORTE_GRAMA: 'CORTE_GRAMA',
  PODA: 'PODA',
  PAISAGISMO: 'PAISAGISMO',
  LIMPEZA: 'LIMPEZA',
  RETIRADA_FOLHAS: 'RETIRADA_FOLHAS',
  ADUBACAO: 'ADUBACAO',
  PLANTIO: 'PLANTIO',
  CONTROLE_PRAGAS: 'CONTROLE_PRAGAS',
  IRRIGACAO: 'IRRIGACAO',
  JARDIM_COMPLETO: 'JARDIM_COMPLETO',
  OUTRO: 'OUTRO',
} as const;
export type ServiceType = (typeof ServiceType)[keyof typeof ServiceType];

/** Human-readable pt-BR labels for the service catalogue. */
export const ServiceTypeLabel: Record<ServiceType, string> = {
  CORTE_GRAMA: 'Cortar grama',
  PODA: 'Poda',
  PAISAGISMO: 'Paisagismo',
  LIMPEZA: 'Limpeza',
  RETIRADA_FOLHAS: 'Retirada de folhas',
  ADUBACAO: 'Adubação',
  PLANTIO: 'Plantio',
  CONTROLE_PRAGAS: 'Controle de pragas',
  IRRIGACAO: 'Sistema de irrigação',
  JARDIM_COMPLETO: 'Jardim completo',
  OUTRO: 'Outro',
};

/**
 * Lifecycle of a service request — an Uber-style state machine.
 * See `docs/14-functional-spec.md` for the allowed transitions.
 */
export const JobStatus = {
  DRAFT: 'DRAFT', // client is still uploading media
  ANALYZING: 'ANALYZING', // AI vision running
  QUOTED: 'QUOTED', // AI estimate ready, awaiting client publish
  MATCHING: 'MATCHING', // published to marketplace, gathering offers
  OFFERED: 'OFFERED', // one or more gardener offers received
  ACCEPTED: 'ACCEPTED', // client chose a gardener
  SCHEDULED: 'SCHEDULED', // date/time agreed
  ENROUTE: 'ENROUTE', // gardener on the way (live tracking)
  ARRIVED: 'ARRIVED', // gardener checked in on-site
  IN_PROGRESS: 'IN_PROGRESS', // work started
  COMPLETED: 'COMPLETED', // gardener checked out, awaiting client approval
  APPROVED: 'APPROVED', // client approved -> payment captured
  PAID: 'PAID', // funds captured and split
  REVIEWED: 'REVIEWED', // rated by client
  CANCELLED: 'CANCELLED',
  DISPUTED: 'DISPUTED',
} as const;
export type JobStatus = (typeof JobStatus)[keyof typeof JobStatus];

/** Status of a gardener's offer against a marketplace job. */
export const OfferStatus = {
  PENDING: 'PENDING',
  ACCEPTED_BY_GARDENER: 'ACCEPTED_BY_GARDENER', // gardener accepted AI price as-is
  COUNTERED: 'COUNTERED', // gardener sent a different price
  DECLINED: 'DECLINED',
  CHOSEN: 'CHOSEN', // client selected this offer
  EXPIRED: 'EXPIRED',
  WITHDRAWN: 'WITHDRAWN',
} as const;
export type OfferStatus = (typeof OfferStatus)[keyof typeof OfferStatus];

/** Payment lifecycle (marketplace escrow with split). */
export const PaymentStatus = {
  PENDING: 'PENDING',
  AUTHORIZED: 'AUTHORIZED', // held (pre-auth / escrow)
  CAPTURED: 'CAPTURED',
  SPLIT: 'SPLIT', // funds distributed to gardener + platform
  REFUNDED: 'REFUNDED',
  FAILED: 'FAILED',
} as const;
export type PaymentStatus = (typeof PaymentStatus)[keyof typeof PaymentStatus];

export const PaymentMethod = {
  PIX: 'PIX',
  CREDIT_CARD: 'CREDIT_CARD',
  GOOGLE_PAY: 'GOOGLE_PAY',
  APPLE_PAY: 'APPLE_PAY',
} as const;
export type PaymentMethod = (typeof PaymentMethod)[keyof typeof PaymentMethod];

export const UserRole = {
  CLIENT: 'CLIENT',
  GARDENER: 'GARDENER',
  ADMIN: 'ADMIN',
  SUPPORT: 'SUPPORT',
} as const;
export type UserRole = (typeof UserRole)[keyof typeof UserRole];

/** Equipment a gardener may own / a job may require. */
export const Equipment = {
  ROCADEIRA: 'ROCADEIRA', // brush cutter
  CORTADOR_GRAMA: 'CORTADOR_GRAMA', // lawn mower
  MOTOSSERRA: 'MOTOSSERRA', // chainsaw
  SOPRADOR: 'SOPRADOR', // leaf blower
  TRITURADOR: 'TRITURADOR', // chipper/shredder
  ESCADA: 'ESCADA', // ladder
  CAMINHAO: 'CAMINHAO', // truck (for hauling)
  PULVERIZADOR: 'PULVERIZADOR', // sprayer
  PODADOR_ALTURA: 'PODADOR_ALTURA', // pole pruner
} as const;
export type Equipment = (typeof Equipment)[keyof typeof Equipment];

export const DifficultyLevel = {
  LOW: 'LOW',
  MEDIUM: 'MEDIUM',
  HIGH: 'HIGH',
  EXTREME: 'EXTREME',
} as const;
export type DifficultyLevel = (typeof DifficultyLevel)[keyof typeof DifficultyLevel];

export const UrgencyLevel = {
  FLEXIBLE: 'FLEXIBLE', // whenever
  NORMAL: 'NORMAL', // within a few days
  URGENT: 'URGENT', // today/tomorrow
  EMERGENCY: 'EMERGENCY', // ASAP
} as const;
export type UrgencyLevel = (typeof UrgencyLevel)[keyof typeof UrgencyLevel];

export const RiskLevel = {
  LOW: 'LOW',
  MEDIUM: 'MEDIUM',
  HIGH: 'HIGH',
} as const;
export type RiskLevel = (typeof RiskLevel)[keyof typeof RiskLevel];

export const TerrainSlope = {
  FLAT: 'FLAT',
  GENTLE: 'GENTLE',
  MODERATE: 'MODERATE',
  STEEP: 'STEEP',
} as const;
export type TerrainSlope = (typeof TerrainSlope)[keyof typeof TerrainSlope];

export const AccessDifficulty = {
  EASY: 'EASY', // street-level, wide gate
  MODERATE: 'MODERATE',
  HARD: 'HARD', // stairs, narrow, no vehicle access
} as const;
export type AccessDifficulty = (typeof AccessDifficulty)[keyof typeof AccessDifficulty];

/** Which vision model produced an analysis (for consensus auditing). */
export const AiProviderId = {
  OPENAI: 'openai',
  GEMINI: 'gemini',
  ANTHROPIC: 'anthropic',
  MOCK: 'mock',
} as const;
export type AiProviderId = (typeof AiProviderId)[keyof typeof AiProviderId];

export const MediaKind = {
  PHOTO: 'PHOTO',
  VIDEO: 'VIDEO',
  AUDIO: 'AUDIO',
  DOCUMENT: 'DOCUMENT',
} as const;
export type MediaKind = (typeof MediaKind)[keyof typeof MediaKind];

/** Kinds of chat message (WhatsApp-style thread between client and gardener). */
export const MessageKind = {
  TEXT: 'TEXT',
  PHOTO: 'PHOTO',
  VIDEO: 'VIDEO',
  AUDIO: 'AUDIO',
  LOCATION: 'LOCATION',
  DOCUMENT: 'DOCUMENT',
  SYSTEM: 'SYSTEM',
} as const;
export type MessageKind = (typeof MessageKind)[keyof typeof MessageKind];
